doc/designs/lms.md

```diff
@@ -0,0 +1,243 @@
+  Due to the problematic nature of managing the stateful OTS private keys,
+  [2] requires that key and signature generation be performed in
+  hardware cryptographic modules that do not allow secret keying material to be
+  exported, even in encrypted form.
+- LMS and LM_OTS signatures.
+  The code will be written in a way that would allow LMS signatures to be added
+  if required, but a HSS signature with only a single level tree should be
+  sufficient for this use case. A separate LMS signature dispatch table could be
+  added if this is ever required.
+- CMS and X509 support - this may be added at a later date.
+- Implementing Extended Merkle signature scheme (XMSS) is a separate effort.
+
+HSS Signature and Key formats
+-----------------------------
+
+The LMS/HSS signature and public key formats use XDR (not ASN1 data).
+The reference code in RFC 8554 checks sizes of the binary data before loading
+the known fields. Rather than do this the OpenSSL PACKET interface is used.
+When the end of the data is expected PACKET_remaining() is tested.
+NOTE that the size of signatures depends on the modes which are embedded in the
+loaded data.
+Byte array data such as encoded public keys are just kept as pointers into the
+data. The entire signature blob is duplicated if the pointers are required
+across API calls.
+
+HSS message streaming API
+-------------------------
+
+In order to handle HSS message streaming the signature data must be decoded
+first, since the message is only used in the last part of the verify.
+This means that the existing sequence of calls for streaming (i.e.
+DigestVerifyInit/DigestVerifyUpdate/DigestFinal) is not sufficient for this
+purpose unless the message is buffered during the update, and then only
+processed during the final.
+
+So we should:
+
+- Support one shot operations:
+  Using existing EVP_DigestVerifyInit_ex + EVP_DigestVerify calls
+- Dont support streaming via existing interface:
+  EVP_DigestVerifyInit_ex + EVP_DigestVerifyUpdate + EVP_DigestVerifyFinal.
+  It could be done, but requires buffering of the message during updates.
+- EVP_VerifyInit etc should not be supported.
+- In order to set the signature early one of the following methods should be
+  used:
+
+Proposal 1:
+
+Add streaming via a new API that is similar to the existing DigestVerify API's,
+but moves the signature |sig| from the final to the init. i.e.
+
+```c
+__owur int EVP_DigestVerifyHBSInit(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
+                                   const char *mdname,
+                                   OSSL_LIB_CTX *libctx,
+                                   const char *props, EVP_PKEY *pkey,
+                                   const OSSL_PARAM params[],
+                                   const unsigned char *sig,
+                                   size_t siglen);   
+/* Alias: since the function signature is identical */
+#define EVP_DigestVerifyHBSUpdate(a,b,c) EVP_DigestVerifyUpdate(a,b,c)
```

Comment:
I know we've done this before - but I think this is very confusing and causes potential problems in the future if we ever want them to do different things.